virttest.qemu_vm: Fix reboot when used with migration

ldoktor opened the pull request:
Running `vm.reboot` while migrating the machine occasionally results in `virttest.qemu_monitor.MonitorSocketError: Could not receive data from monitor    ([Errno 104] Connection reset by peer)`. This is caused by using QMP monitor of the source (dead) VM.

While fixing that I also noticed changed behavior after 3c2f6241a that always uses QMP go_down method when available, even though the `shell` method defines it's own check_down method (that in fact does not suffer the migration issue). The first commits changes the logic to only use QMP when no better method is defined.

ldoktor commented:
Note I tested this with debugger, but it is also possible to reproduce using SIGSTOP or VM that takes long to reboot.

luckyh commented:
I can't recall too many details, but I remember that the problem is the current implementation of `_go_down()` is not so robust, sometimes we could get false-positive results from that, so, that is why I introduced the event checker and made it as the default.

luckyh commented on virttest/qemu_vm.py:
```diff
@@ -4266,7 +4266,20 @@ def _shell_reboot(session, timeout):
+                original_monitor = self.monitor
+
+                def _check_go_down():
+                    if self.monitor != original_monitor:
```
How about:
```python
original_monitor = id(self.monitor)
...
    if id(self.monitor) != original_monitor:
        ...
```

luckyh commented on virttest/qemu_vm.py:
```diff
@@ -4266,7 +4266,20 @@ def _shell_reboot(session, timeout):
+                                        "reboot, checking for reboot is not "
+                                        "reliable. Assuming it already "
+                                        "rebooted.")
+                        return True
```
Here we have two choices:

1. always return `True` by monitor change, then getting false-positive results is the side effect.
2. continue looking up the event from destination side, then getting false-negative results is the side effect (in this case, the event has been sent to source side but we missed that).

I prefer the second one because it would less likely give the wrong results than the first one.

ldoktor replied:
Actually I was trying it with a VM that took long to reboot and almost always it rebooted in the wrong time. That is why I chose to return `True`. Note that we also return True on human monitor by default so it's nothing uncommon.

Anyway even if we get false-positive, VM should not accept new connections while rebooting, therefor we should get failed-to-boot. I mean currently about 10% of my jobs are failing because of this, changing it to default False it would be something between 1-10% so it should be better but honestly I'd prefer wrong failed message and ~0% of failures than 1-10% of failures where I'd have to look whether it was because of missed qmp event or the machine actually refusing to go down.

So the question is, do you just slightly prefer it, or is it a strong nack? It should not introduce false-positives, it could just occasionally report failed-to-boot instead of failed-to-shutdown-before-boot.

luckyh replied:
Sorry for the delayed response.

> Anyway even if we get false-positive, VM should not accept new connections while rebooting, therefor we should get failed-to-boot.

Well, "failed-to-boot" (I am not sure about what you mean by this error, so I understand it as a kind of `Exception`) is just one of the moderate errors, I mean at least the function would raise exceptions in that case. The worst thing is the function would successfully return with a new session while the guest is just at the beginning to commit a reboot, if the sequence goes like as the follow it could happen:

```
 time \ routines |           vm.reboot()          | migration |        os system service        |
-----------------+--------------------------------+-----------+---------------------------------+
        t1       |   _reboot() (shell) returned   |           |                                 |
        t2       |                                | completed | preparing to shut down services |
        t3       | _check_go_down() returned True |           |                                 |
        t4       |   calling vm.wait_for_login()  |           |                                 |
        t5       |        reboot() returned       |           |        shutting down sshd       |
        t6       |                                |           |          shut down sshd         |
        t7       |                                |           |            rebooting            |
```

And I guess we may meet this more in case of rdma migration.

> I mean currently about 10% of my jobs are failing because of this, changing it to default False it would be something between 1-10% so it should be better but honestly I'd prefer wrong failed message and ~0% of failures than 1-10% of failures where I'd have to look whether it was because of missed qmp event or the machine actually refusing to go down.
> 
> So the question is, do you just slightly prefer it, or is it a strong nack? It should not introduce false-positives, it could just occasionally report failed-to-boot instead of failed-to-shutdown-before-boot.


Actually, I do want to request the changes, but I'd like to know your opinions at first.

Thanks.

ldoktor replied:
I'd call that unlikely, but I'm not strongly tied to this solution and I understand your concern. Let me simply keep using the current monitor, which means we can miss the event, but we should not declare the reboot complete unless the RESET event is emitted.

luckyh commented:
I wrote a patch at before but have not sent it out yet, just for your information.
```patch
diff --git a/virttest/qemu_vm.py b/virttest/qemu_vm.py
index e622dee0..83313c14 100644
--- a/virttest/qemu_vm.py
+++ b/virttest/qemu_vm.py
@@ -4159,6 +4159,27 @@ class VM(virt_vm.BaseVM):
             except Exception:
                 return True
 
+        def _go_down_qmp():
+            # XXX we may at here during migration, so if it happens just
+            # right before migration completed, an `MonitorSocketError`
+            # would be emitted because the source end qemu process had
+            # quitted. We could just ignore the error and continue to
+            # capture the event we want at the destination end.
+            #
+            # This is just a workaround of the issue, because the thing
+            # should be handled by `monitor.get_event()` well. But
+            # unfortunately it requires to refactor whole `qemu_monitor`
+            # approach that takes lots of time, leave this stuff to the
+            # future.
+            #
+            # And also, since there is a race condition to be solved,
+            # you might hit `VMRebootError`, so please don't be surprised
+            # of that. :)
+            try:
+                return bool(self.monitor.get_event("RESET"))
+            except qemu_monitor.MonitorSocketError:
+                return False
+
         def _shell_reboot(session, timeout):
             if not session:
                 if not serial:
@@ -4185,7 +4206,7 @@ class VM(virt_vm.BaseVM):
             raise virt_vm.VMRebootError("Unknown reboot method: %s" % method)
 
         if isinstance(self.monitor, qemu_monitor.QMPMonitor):
-            _check_go_down = partial(self.monitor.get_event, "RESET")
+            _check_go_down = _go_down_qmp
             self.monitor.clear_event("RESET")
 
         try:
```

ldoktor commented:
@luckyh I changed the second commit to always use the current monitor command and added warning. This situation is not that common so I log it every time as it doesn't seem worth complicating it to log it just once.

luckyh commented:
> @luckyh I changed the second commit to always use the current monitor command and added warning. This situation is not that common so I log it every time as it doesn't seem worth complicating it to log it just once.

ok, thanks. Then the only issue is, as I mentioned in the first comment, IIRC, `_go_down` may give false-positive as well, so if we have the first commit we might get regressions. I'm not sure about how likely it would be, but I think we can have it at first, and if it does make regressions then we can revert it accordingly, how about that?

ldoktor commented:
It's hard to judge which method is more accurate. In my testing it was `_go_down`, but I do want to get this merged so let's use QMP always when available.

yanan-fu commented:
Ack

luckyh commented:
Thanks all.
